lib/tty/prompt/multi_list.rb

```diff
@@ -60,7 +60,7 @@ def keyspace(*)
         else
           return if @max && @selected.size >= @max
           @selected << active_choice
-          @selected.sort_by! { |choice| @choices.index(choice) }
+          @selected = @choices.to_ary & @selected
```

Comment:
@piotrmurach look, the simplest trick ever as `@choices` IS our index cache :-)

Reply:
Nice one Daniel! I used a similar set operation technique for the `keyctrl_r` implementation. One suggestion I would have is to make the `&` a delegate method so we can skip the `to_ary` call. This may speed things up even further.

Reply:
Right, good idea, will do :+1:

Reply:
Added a [benchmark file](https://github.com/piotrmurach/tty-prompt/blob/master/benchmarks/multi_select.rb) for multi_select prompt that simulates selection of 100 first items from 10,000 choices. This should give us a common ground for improvements/experimentation. From my benchmark runs(per 10 seconds), I get:

```
48 times - baseline
37 times - sort_by!  - 25% reduction in speed
13 times - array & - 75% reduction in speed 
47 times - parallel indexes - similar perf to baseline
```

Unfortunately, what you suggested seems actually slower than sorting in place. I guess that kind of makes sense given that each time a choice is selected, an intersection operation is done with a large collection of 10,000 items. I had a quick look at trying to improve performance and come up with 'parallel indexes' solution.

The idea behind 'parallel indexes' is to keep the insertion order for selected choices in a separate array. We already know the active index, so the only thing to figure out is the insertion place in the indexes array - this could be a binary search. Once this is done, we can insert a new index into indexes array and active choice into the selected array. This way we constantly keep inserting things in order in O(1) time. This doesn't 'suffer' from the need to look through `@choices` that may be really large. Similarly, it doesn't require sorting on each addition.

As you suggested before, this could be a new structure `SelectedChoices` that handles internally the ordering of items. It would only need to have knowledge of the active index and a current choice. Plus as I was looking at improving things the `@selected.size` is called repeatedly in few places. Therefore, we could also consider caching the `@size` as well.

What do you reckon?

Reply:
Hum, that thing about speeds is strange, I'll double check why my benchmark is so radically different. Also good idea about caching `#size` as well as indices, I'll try it out, will let you know here